---
title: "Peak heatmap"
author: "Bing Shui"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: yes
    toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Prepare libraies
```{r message=FALSE, warning=FALSE}
library(data.table)
library(tidyverse)
library(rtracklayer)
library(DESeq2)
library(Biostrings)
library(genomation)
library(RColorBrewer)
library(CLIPanalyze)
```

## Calculate scaleFactor for `bamCoverage`
`bamCoverage` from [Deeptools](https://deeptools.readthedocs.io/en/develop/content/tools/bamCoverage.html) was used to generated scaled `bw` files for plotting peak heatmaps

Scalefactors are calculated as the reciprical of DESeq2 estimated sizeFactors using counnts in genes outside of peaks.

```{r}
dir.create("PDF_figure/Peak_heatmap_merged", showWarnings = FALSE)

load('../Merged_Analysis/merged_peak_analysis.rda')

sizefactor <- sizeFactors(dds.peaks.hva.hvak)

scalefactor <- (1/sizefactor)/max((1/sizefactor))

scalefactor
```

`bw` file generation was done on cluster.

## Load miR-peaks dataset
```{r}
mirs.peaks <- readRDS("Datafiles/miRNA-merged-peaks-list-09282019.rds")
```

## Analysis
Prepare bw file for visualization
```{bash eval = F}
# combine bigwig files of the same genotype
$ cd ../Rescaled analysis/HVA
$ bigWigMerge *.bw HVA_merged.bedGraph

$ cd ../Rescaled analysis/HVAK
$ bigWigMerge *.bw HVAK_merged.bedGraph
```

convert chromosome name from Ensembl to UCSC format.
```{r eval = F}
chromosomename <- read.table("../Rescaled analysis/Data/rescaled_bw_deseq_sizefactor/Ensembl.mm10.chrom.sizes.txt", sep = "\t")[,1][1:22]
hva_bed <- read.table("../Rescaled analysis/Data/rescaled_bw_deseq_sizefactor/HVA/HVA_merged.bedGraph", sep = "\t")
hva_bed <- hva_bed %>% filter(V1 %in% chromosomename)
hva_bed$V1 <- paste0("chr", hva_bed$V1)
hva_bed$V1[hva_bed$V1 == "chrMT"] <- "chrM"
write_delim(hva_bed, file = "../Rescaled analysis/Data/rescaled_bw_deseq_sizefactor/HVA_merged_edited.bedGraph", delim = "\t", col_names = F)

hvak_bed <- read.table("../Rescaled analysis/Data/rescaled_bw_deseq_sizefactor/HVAK/HVAK_merged.bedGraph", sep = "\t")
hvak_bed <- hvak_bed %>% filter(V1 %in% chromosomename)
hvak_bed$V1 <- paste0("chr", hvak_bed$V1)
hvak_bed$V1[hvak_bed$V1 == "chrMT"] <- "chrM"
write_delim(hvak_bed, file = "../Rescaled analysis/Data/rescaled_bw_deseq_sizefactor/HVAK_merged_edited.bedGraph", delim = "\t", col_names = F)

```

```{bash eval = F}
# bedGraph to bigWig conversion
$ awk 'NR!=1' HVA_merged_edited.bedGraph > input.HVA_merged.bedGraph
$ sort -k1,1 -k2,2n input.HVA_merged.bedGraph > sorted.input.HVA_merged.bedGraph
$ bedGraphToBigWig sorted.input.HVA_merged.bedGraph mm10.chrom.sizes HVA_merged.bw

$ awk 'NR!=1' HVAK_merged_edited.bedGraph > input.HVAK_merged.bedGraph
$ sort -k1,1 -k2,2n input.HVAK_merged.bedGraph > sorted.input.HVAK_merged.bedGraph
$ bedGraphToBigWig sorted.input.HVAK_merged.bedGraph mm10.chrom.sizes HVAK_merged.bw
```

```{r}
peaks <- mirs.peaks
peaks <- lapply(peaks, FUN = function(x) {x[order(x$count, decreasing = T)]})
mybw.dir <- "../Rescaled analysis/Data/rescaled_bw_deseq_sizefactor"
mybw.files <- list.files(mybw.dir, pattern = "bw$", full.names = T)
print(mybw.files)

cols <- brewer.pal(name = "Set2", n = 8)
reds <- brewer.pal(name = "Reds", n = 9)
blues <- brewer.pal(name = "Blues", n = 9)
mycolors <-c(blues[6], reds[6]) #HVA, HVAK
light.colors <- alpha(c(blues[3], reds[2]), 0.4)
```

### Histogram
histogram plot function
```{r}
#plot histograms
peaks_meta <- function(mypeaks = peaks, 
                       miRNA_family = "miR-451a", 
                       dispersion = "se",
                       dispersion.col = NULL,
                       coordinates = c(-400, 400), 
                       line.col = mycolors, 
                       winsorize = c(0,99),
                       title = ""){
  
  suppressWarnings(mypeaks <- lapply(mypeaks, FUN = function(x) {resize(x, width = sum(abs(coordinates)), fix="center")}))
  mypeaks <- GRangesList(mypeaks)
  mysml <- ScoreMatrixList(targets=mybw.files, window=mypeaks[[miRNA_family]], type = "bigWig", strand.aware = T)
  mysampleInfo <- data.frame(basename(mybw.files), c("HVA", "HVAK"))
  names(mysampleInfo) = c("sample", "genotype")
  names(mysml) = mysampleInfo$genotype[match(names(mysml), mysampleInfo$sample)]
  plotMeta(mysml, profile.names = c("HVA", "HVAK"), xcoords = coordinates, dispersion = dispersion, main =title, line.col = line.col, winsorize = winsorize, dispersion.col = dispersion.col)
}
```

```{r}
len <- sapply(mirs.peaks, function(x) length(x))
mirs.peaks <- mirs.peaks[order(-len)]
mirna <- names(mirs.peaks)

for (i in 1:10) {
  peaks_meta(mypeaks = peaks, dispersion  = NULL, miRNA_family = mirna[i], title = mirna[i])
  peaks_meta(mypeaks = peaks, dispersion  = "se", miRNA_family = mirna[i], title = mirna[i],
           dispersion.col = light.colors)
}

pdf("PDF_figure/Peak_heatmap_merged//Histogram_peak.pdf",
    height = 4,
    width = 6)
for (i in 1:10) {
  peaks_meta(mypeaks = peaks, dispersion  = NULL, miRNA_family = mirna[i], title = mirna[i])
  peaks_meta(mypeaks = peaks, dispersion  = "se", miRNA_family = mirna[i], title = mirna[i],
           dispersion.col = light.colors)
}
dev.off()
```

### Heatmap
plot heatmaps
```{r}
peaks_heat <- function(mypeaks = peaks, 
                       miRNA_family = "miR-451", 
                       col = blues9, 
                       coordinates = c(-400, 400), 
                       order_rows = F, 
                       winsorize_parameters = c(1,98)){
  suppressWarnings(mypeaks <- lapply(mypeaks, FUN = function(x) {resize(x, width = sum(abs(coordinates)), fix="center")}))
  mypeaks <- GRangesList(mypeaks)
  mysml <- ScoreMatrixList(targets=mybw.files[c(1,2)], window=mypeaks[[miRNA_family]], type = "bigWig", strand.aware = T)
  mysampleInfo <- data.frame(basename(mybw.files), c("HVA", "HVAK"))
  names(mysampleInfo) = c("sample", "genotype")
  names(mysml) = paste(mysampleInfo$genotype[match(names(mysml), mysampleInfo$sample)], miRNA_family , sep = "_")
  mysml.scaled = scaleScoreMatrixList(mysml)
  #multiHeatMatrix(mysml.scaled, xcoords = coordinates, col = col)
  multiHeatMatrix(mysml, common.scale = T, xcoords = coordinates, winsorize = winsorize_parameters, col = col, order = order_rows)
}
```

```{r}
colfunc <- colorRampPalette(c("white", "blue"))
mycols <- colfunc(128)
```

```{r fig.height=8, fig.width=8}
for (i in 1:10) {
  peaks_heat(mypeaks = peaks, miRNA_family = mirna[i], col = blues9, winsorize_parameters = c(0,99))
}

pdf("PDF_figure/Peak_heatmap_merged//Heatmap_peak.pdf",
    height = 8,
    width = 8)
for (i in 1:10) {
  peaks_heat(mypeaks = peaks, miRNA_family = mirna[i], col = blues9, winsorize_parameters = c(0,99))
}
dev.off()
```

### SessionInfo
```{r}
sessionInfo()
```
