---
title: "Up_down-regualted_HEAP_peaks"
author: "Bing Shui"
date: "12/3/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Library Loading for the analysis
```{r}
suppressMessages(
suppressWarnings(
  c(library(data.table),
    library(tidyverse),
    library(rtracklayer),
    library(Rsamtools),
    library(DESeq2),
    library(Rsubread),
    library(CLIPanalyze),
    library(RColorBrewer),
    library(gplots),
    library(Biostrings))
  )
)
```

## Load the peakdata
```{r}
filter.peaks<- readRDS("Datafiles/peaks-filtered-09282019.rds")
```

## Kmer enrichment analysis
### Up-regulated peaks
```{r}
padj.threshold <- 0.05
up.peaks <- subset(filter.peaks, filter.peaks$hvak.hva.log2FC > 0 & filter.peaks$hvak.hva.padj < padj.threshold)
```

Load miRBase mature miRNA fa file and mm10 genome
```{r}
mirnas <- readRNAStringSet("/Users/mizuhi/OneDrive - Harvard University/Haigis Lab/Projects/Halo-Ago2/Halo-Ago-KRas/Raw Data/HEAP-CLIP/HEAP_09282019/Analysis/CLIP_Analysis/Data Visualization/mature.fa")
mirnas <- DNAStringSet(mirnas[grepl("mmu-", names(mirnas))])
names(mirnas) <- sapply(strsplit(names(mirnas), " "), "[", 1)
names(mirnas) <- substring(names(mirnas), first = 5)
bsgenome <- load.bsgenome("mm10")
```

```{r}
# define functions
associateKmerWithMiRNA <- function(kmer, mirnas, in.seed = TRUE,
                                   collapse = TRUE) {
  occurrences <- unlist(vmatchPattern(reverseComplement(DNAString(kmer)),
                                      mirnas))
  if (in.seed) {
    occurrences <- occurrences[end(occurrences) <= 8]
  }
  mirna.names <- sort(names(occurrences))
  if (collapse) {
    mirna.names <- paste0(mirna.names, collapse = ",")
  }
  return(mirna.names)
}

findEnrichedKmersPeaks <- function(peaks, kmer.background, k = 6, n = 50) {
  enriched.kmers <- findKmerEnrich(peaks, k = k, genomeTag = "mm10",
                                   kmer.background = kmer.background)[1:n]
  enriched.kmers <-
    data.table(kmer = names(enriched.kmers),
               enrich = enriched.kmers,
               miR = sapply(names(enriched.kmers),
                            associateKmerWithMiRNA, mirnas = mirnas))
  enriched.kmers
}

# establish proper background
# Background are filter.peaks
# calculate all the kmer backgrounds
k6.background.overlap <- calculateKmerBackground(k=6, genomeTag = "mm10", gr.overlap = filter.peaks,
                                         exons.only = TRUE)
k7.background.overlap <- calculateKmerBackground(k=7, genomeTag = "mm10", gr.overlap = filter.peaks,
                                         exons.only = TRUE)
k8.background.overlap <- calculateKmerBackground(k=8, genomeTag = "mm10", gr.overlap = filter.peaks,
                                         exons.only = TRUE)

# 6mer
up.peaks.k6 <- findEnrichedKmersPeaks(up.peaks, k6.background.overlap, k=6)
ggplot(up.peaks.k6[1:10, ],
       aes(x = reorder(kmer, enrich), y = enrich)) +
  geom_col() + coord_flip() + theme_bw() + xlab("") +
  ylab("Frequency\nlog2(peaks / background)") +
  labs(title = "Enriched 6mers in upregulated peaks")
up.peaks.k6[1:10,]

# 7mers
up.peaks.k7 <- findEnrichedKmersPeaks(up.peaks, k7.background.overlap, k=7)
ggplot(up.peaks.k7[1:10, ],
       aes(x = reorder(kmer, enrich), y = enrich)) +
  geom_col() + coord_flip() + theme_bw() + xlab("") +
  ylab("Frequency\nlog2(peaks / background)") +
  labs(title = "Enriched 7mers in upregulated peaks")
up.peaks.k7[1:10,]

# 8mers
up.peaks.k8 <- findEnrichedKmersPeaks(up.peaks, k8.background.overlap, k=8)
ggplot(up.peaks.k8[1:10, ],
       aes(x = reorder(kmer, enrich), y = enrich)) +
  geom_col() + coord_flip() + theme_bw() + xlab("") +
  ylab("Frequency\nlog2(peaks / background)") +
  labs(title = "Enriched 8mers in upregulatedd peaks")
up.peaks.k8[1:10,]
```

### Down-regulated peaks
```{r}
down.peaks <- subset(filter.peaks, filter.peaks$hvak.hva.log2FC < 0 & filter.peaks$hvak.hva.padj < padj.threshold)
```

```{r}
# 6mer
down.peaks.k6 <- findEnrichedKmersPeaks(down.peaks, k6.background.overlap, k=6)
ggplot(down.peaks.k6[1:10, ],
       aes(x = reorder(kmer, enrich), y = enrich)) +
  geom_col() + coord_flip() + theme_bw() + xlab("") +
  ylab("Frequency\nlog2(peaks / background)") +
  labs(title = "Enriched 6mers in downregulated peaks")
down.peaks.k6[1:10,]

# 7mers
down.peaks.k7 <- findEnrichedKmersPeaks(down.peaks, k7.background.overlap, k=7)
ggplot(down.peaks.k7[1:10, ],
       aes(x = reorder(kmer, enrich), y = enrich)) +
  geom_col() + coord_flip() + theme_bw() + xlab("") +
  ylab("Frequency\nlog2(peaks / background)") +
  labs(title = "Enriched 7mers in downregulated peaks")
down.peaks.k7[1:10,]

# 8mers
down.peaks.k8 <- findEnrichedKmersPeaks(down.peaks, k8.background.overlap, k=8)
ggplot(down.peaks.k8[1:10, ],
       aes(x = reorder(kmer, enrich), y = enrich)) +
  geom_col() + coord_flip() + theme_bw() + xlab("") +
  ylab("Frequency\nlog2(peaks / background)") +
  labs(title = "Enriched 8mers in downregulated peaks")
down.peaks.k8[1:10,]
```

As a sanity check, we need to compare the above enrichment with enrichment of k-mers in whole transcript sequences.
```{r}
findEnrichedKmersSeqs <- function(seqs, kmer.background, k = 6, n = 50) {
  enriched.kmers <- findKmerEnrich(gr.seq = seqs, k = k, genomeTag = "mm10",
                                   kmer.background = kmer.background)[1:n]
  enriched.kmers <-
    data.table(kmer = names(enriched.kmers),
               enrich = enriched.kmers,
               miR = sapply(names(enriched.kmers),
                            associateKmerWithMiRNA, mirnas = mirnas))
  enriched.kmers
}

mm10.annot <- loadAnnot("mm10")
mm10.transcript <- transcriptsBy(mm10.annot$txdb)
mm10.transcript <- unlist(mm10.transcript)
mm10.withpeaks <- subsetByOverlaps(mm10.transcript, filter.peaks)
mm10.withpeaks.seq <- get.seqs(bsgenome, mm10.withpeaks)

# 6mer
transcript.overlap.k6 <- findEnrichedKmersSeqs(mm10.withpeaks.seq, k6.background.overlap)
ggplot(transcript.overlap.k6[1:10, ],
       aes(x = reorder(kmer, enrich), y = enrich)) +
  geom_col() + coord_flip() + theme_bw() + xlab("") +
  ylab("Frequency\nlog2(transcript / background)") +
  labs(title = "Enriched 6mers in all transcripts with peaks")
transcript.overlap.k6[1:10, ]

# 7mer
transcript.overlap.k7 <- findEnrichedKmersSeqs(mm10.withpeaks.seq, k7.background.overlap, k=7)
ggplot(transcript.overlap.k7[1:10, ],
       aes(x = reorder(kmer, enrich), y = enrich)) +
  geom_col() + coord_flip() + theme_bw() + xlab("") +
  ylab("Frequency\nlog2(transcript / background)") +
  labs(title = "Enriched 7mers in all transcripts with peaks")
transcript.overlap.k7[1:10, ]

# 8mer
transcript.overlap.k8 <- findEnrichedKmersSeqs(mm10.withpeaks.seq, k8.background.overlap, k=8)
ggplot(transcript.overlap.k8[1:10, ],
       aes(x = reorder(kmer, enrich), y = enrich)) +
  geom_col() + coord_flip() + theme_bw() + xlab("") +
  ylab("Frequency\nlog2(transcript / background)") +
  labs(title = "Enriched 8mers in all transcripts with peaks")
transcript.overlap.k8[1:10, ]
```

## CAD analysis
CAD analysis will have to be done using specific miRNAs since it would be pointless to do just HVA or HVAK.

Set up the function
```{r}
plotCDF.ggplot <- function(gene.counts, gene.sets, gene.set.labels,
                           col = "", linetype = "", xlim = c( -1.0, 1.3 ),
                           legend.size = 22, axistitle.size = 22, title = "Fold change log2 (Dicer KO/WT)",
                           legend.pos = c(0.7, 0.18)) {
  require(ggplot2)
  df.log2fc <- gene.counts[,c("gene", "log2FC")]
  #rownames(df.log2fc) <- df.log2fc$gene
  if (length(gene.sets) != length(gene.set.labels)){
    return("Length of gene sets doesn't match labels")
  }
  target.expr <- df.log2fc[df.log2fc$gene %in% gene.sets[[1]],]
  for (i in 2:length(gene.sets)){
    target.expr <- rbind(target.expr, df.log2fc[df.log2fc$gene %in% gene.sets[[i]],])
  }

  gene.set.counts <- c()
  for (j in 1:length(gene.sets)){
    gene.set.counts <- c(gene.set.counts, sum(df.log2fc$gene %in% gene.sets[[j]]))
  }
  
  target.expr$Category <- rep(gene.set.labels, gene.set.counts)
  target.expr$Category <- factor(target.expr$Category, levels = gene.set.labels)
  
  log2FC.values <- lapply(gene.sets, function(gene.set) {
    gene.counts[gene.counts$gene %in% gene.set,]$log2FC
  })

  ks.pvals <- lapply(log2FC.values,
                     function(log2FCs) {
                       ks.test(log2FCs, log2FC.values[[1]])$p.value
                     })
  
  p <- ggplot( target.expr, aes( x = log2FC, colour = Category ) ) +
  stat_ecdf( geom = 'step', aes( colour = Category, linetype = Category ), lwd = 1 ) +
  scale_color_manual( values = col, labels = sprintf( "%s (%d)", gene.set.labels, gene.set.counts ) ) +
  scale_linetype_manual( values = linetype, labels = sprintf( "%s (%d)", gene.set.labels, gene.set.counts ) ) +
  # xlim() will remove data points; Be careful in the future
  coord_cartesian( xlim = xlim ) + xlab(title) + ylab('CDF') +
  theme_classic() + theme( legend.background = element_rect(fill = NA), 
                           legend.title = element_blank(), 
                           legend.position = legend.pos,
                           legend.text = element_text(size=legend.size),
                           legend.key.size = unit(1.5, 'lines'),
                           axis.title.x = element_text(size=axistitle.size, margin = margin(t = 10)),
                           axis.title.y = element_text(size=axistitle.size, margin = margin(r = 10)),
                           axis.text=element_text(size=20),
                           axis.line = element_line(size = 1),  #axis label size
                           axis.ticks.length = unit(0.3, "cm")) #increase tick length
  
  for (k in 2:length(gene.sets)){
    p <- p + annotate(geom = "text", x = -0.9, y = 1-0.08*(k-1), hjust = 0, 
                      label = sprintf("p = %.0e", ks.pvals[k]), 
                      colour = col[k], size = 8)
  }
  print(p)
}
```

```{r}
peak.DGE <- as.data.frame(filter.peaks)
colnames(peak.DGE)[7] <- "gene"
colnames(peak.DGE)[18] <- "lfc"
colnames(peak.DGE)[26] <- "log2FC"
# input miRNA list with associated peaks
mirs.peaks <- readRDS("Datafiles/miRNA-peaks-list-09282019-withIDs.rds")
cols <- c(brewer.pal(name = "Set2", n = 8))
```

### miR-29a/b/c-3p peaks
```{r}
plotCDF.ggplot(peak.DGE,
               list(peak.DGE$gene, mirs.peaks$`miR-29-3p`$name),
               c("All peaks", "miR-29-3p"),
               col = c("grey15", cols[1]),
               linetype = c(1, 1),
               title = "CLIP_LFC(HVAK/HVA)",
               xlim = c(-5,4)
               )
```

### miR-194-5p peaks
```{r}
plotCDF.ggplot(peak.DGE,
               list(peak.DGE$gene, mirs.peaks$`miR-194-5p`$name),
               c("All peaks", "miR-194-5p"),
               col = c("grey15", cols[2]),
               linetype = c(1, 1),
               title = "CLIP_LFC(HVAK/HVA)",
               xlim = c(-5,4)
               )
```